pallets/parachain-staking/src/lib.rs

```diff
@@ -0,0 +1,2630 @@
+			Self::do_inc_collator_reward(&collator, before_stake);
+
+			Self::deposit_event(Event::CollatorStakedLess(collator, before_stake, after));
+			Ok(Some(<T as pallet::Config>::WeightInfo::candidate_stake_less(
+				n,
+				T::MaxDelegatorsPerCollator::get(),
+			))
+			.into())
+		}
+
+		/// Join the set of delegators by delegating to a collator candidate.
+		///
+		/// The account that wants to delegate cannot be part of the collator
+		/// candidates set as well.
+		///
+		/// The caller must _not_ have a delegation. If that is the case, they
+		/// are required to first remove the delegation.
+		///
+		/// The amount staked must be larger than the minimum required to become
+		/// a delegator as set in the pallet's configuration.
+		///
+		/// As only `MaxDelegatorsPerCollator` are allowed to delegate a given
+		/// collator, the amount staked must be larger than the lowest one in
+		/// the current set of delegator for the operation to be meaningful.
+		///
+		/// The collator's total stake as well as the pallet's total stake are
+		/// increased accordingly.
+		///
+		/// Emits `Delegation`.
+		/// Emits `DelegationReplaced` if the candidate has
+		/// `MaxDelegatorsPerCollator` many delegations but this delegator
+		/// staked more than one of the other delegators of this candidate.
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::join_delegators(
+			T::MaxTopCandidates::get(),
+			T::MaxDelegatorsPerCollator::get()
+		))]
+		pub fn join_delegators(
+			origin: OriginFor<T>,
+			collator: <T::Lookup as StaticLookup>::Source,
+			amount: BalanceOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let acc = ensure_signed(origin)?;
+			let collator = T::Lookup::lookup(collator)?;
+
+			// check balance
+			ensure!(
+				pallet_balances::Pallet::<T>::free_balance(acc.clone()) >= amount.into(),
+				pallet_balances::Error::<T>::InsufficientBalance
+			);
+
+			// first delegation
+			ensure!(DelegatorState::<T>::get(&acc).is_none(), Error::<T>::AlreadyDelegating);
+			ensure!(amount >= T::MinDelegatorStake::get(), Error::<T>::DelegationBelowMin);
+
+			// cannot be a collator candidate and delegator with same AccountId
+			ensure!(Self::is_active_candidate(&acc).is_none(), Error::<T>::CandidateExists);
+			ensure!(
+				Unstaking::<T>::get(&acc).len().saturated_into::<u32>() < T::MaxUnstakeRequests::get(),
+				Error::<T>::CannotJoinBeforeUnlocking
+			);
+			// cannot delegate if number of delegations in this round exceeds
+			// MaxDelegationsPerRound
+			let delegation_counter = Self::get_delegation_counter(&acc)?;
+
+			// prepare update of collator state
+			let mut state = CandidatePool::<T>::get(&collator).ok_or(Error::<T>::CandidateNotFound)?;
+			let num_delegations_pre_insertion: u32 = state.delegators.len().saturated_into();
+
+			ensure!(!state.is_leaving(), Error::<T>::CannotDelegateIfLeaving);
+			let delegation = Stake {
+				owner: acc.clone(),
+				amount,
+			};
+
+			// attempt to insert delegator and check for uniqueness
```

Comment:
if we specify explicitly two different handling case when MaxDelegatorsPerCollator equal to num_delegations_pre insertion and when not equal then it will looks better for reading and algorithm complexity will stay the same.

let delegator_state = Delegator { amount, owner: Some(collator.clone()) };
let CandidateOf::<T, _> { stake: old_stake, total: old_total, .. } = state;




	let state = if num_delegations_pre_insertion == T::MaxDelegatorsPerCollator::get(){
		ensure!(state.delegators.contains(&delegation.clone()), Error::<T>::DelegatorExists);
		Self::do_update_delegator(delegation, state)?
	}
	else {
		let insert_delegator = state
		.delegators
		.try_insert(delegation.clone()).unwrap();
		ensure!(insert_delegator, Error::<T>::DelegatorExists);
		state.total = state.total.saturating_add(amount);
		state
	};